Source/HTTPHeaders.swift

```diff
@@ -24,51 +24,215 @@
 
 import Foundation
 
-public typealias HTTPHeaders = [String: String]
+public struct HTTPHeaders {
+    private var headers = [HTTPHeader]()
+    
+    public init() { }
+    
+    public init(_ headers: [HTTPHeader]) {
+        self.headers = headers
+    }
+    
+    public init(_ dictionary: [String: String]) {
+        self.init()
+        
+        headers = dictionary.map { HTTPHeader(name: $0.key, value: $0.value) }
+    }
+    
+    public mutating func update(name: String, value: String) {
+        guard let index = headers.index(of: name) else {
+            headers.append(HTTPHeader(name: name, value: value))
+            return
+        }
+        
+        headers.replaceSubrange(index...index, with: [HTTPHeader(name: name, value: value)])
+    }
+    
+    public mutating func update(_ header: HTTPHeader) {
+        update(name: header.name, value: header.value)
+    }
+    
+    public mutating func remove(name: String) {
```

Comment:
To avoid calling `lowercased()` twice for each comparison:

```swift
mutating func remove(name: String) {
    let lowercasedName = name.lowercased()
    guard let index = headers.firstIndex(where: { $0.name.lowercased() == lowercasedName } ) else { return }
    headers.remove(at: index)
}
```

Reply:
Done in the `Array` extension.